runtime/src/main/java/org/corfudb/protocols/CorfuProtocolCommon.java

```diff
@@ -0,0 +1,214 @@
+     *
+     * @param layout   the desired (Java) Layout
+     */
+    public static LayoutMsg getLayoutMsg(Layout layout) {
+        return LayoutMsg.newBuilder()
+                .setLayoutJson(layout.asJSONString())
+                .build();
+    }
+
+    /**
+     * Returns a Layout object from its Protobuf representation.
+     *
+     * @param msg   the Protobuf Layout message
+     */
+    public static Layout getLayout(LayoutMsg msg) {
+        return Layout.fromJSONString(msg.getLayoutJson());
+    }
+
+    /**
+     * Returns the Protobuf representation of a Token, given
+     * its epoch and sequence.
+     *
+     * @param epoch      the epoch of the token
+     * @param sequence   the sequencer number of the token
+     */
+    public static TokenMsg getTokenMsg(long epoch, long sequence) {
+        return TokenMsg.newBuilder()
+                .setEpoch(epoch)
+                .setSequence(sequence)
+                .build();
+    }
+
+    /**
+     * Returns the Protobuf representation of a Token.
+     *
+     * @param token   the desired (Java) Token
+     */
+    public static TokenMsg getTokenMsg(Token token) {
+        return getTokenMsg(token.getEpoch(), token.getSequence());
+    }
+
+    /**
+     * Returns the Protobuf representation of a SequencerMetrics object.
+     *
+     * @param metrics   the desired (Java) SequencerMetrics
+     */
+    public static SequencerMetricsMsg getSequencerMetricsMsg(SequencerMetrics metrics) {
+        return SequencerMetricsMsg.newBuilder()
+                .setSequencerStatus(sequencerStatusTypeMap.getOrDefault(
+                        metrics.getSequencerStatus(), SequencerStatus.INVALID))
+                .build();
+    }
+
+    /**
+     * Returns a SequencerMetrics object from its Protobuf representation.
+     *
+     * @param msg   the Protobuf SequencerMetrics message
+     */
+    public static SequencerMetrics getSequencerMetrics(SequencerMetricsMsg msg) {
+        switch (msg.getSequencerStatus()) {
+            case READY: return SequencerMetrics.READY;
+            case NOT_READY: return SequencerMetrics.NOT_READY;
+            case UNKNOWN: return SequencerMetrics.UNKNOWN;
+            default:
+                throw new UnsupportedOperationException("SequencerMetrics message unrecognized: "
+                        + "Status=" + msg.getSequencerStatus());
+        }
+    }
+
+    /**
+     * Returns the Protobuf representation of a StreamAddressSpace object.
+     *
+     * @param addressSpace   the desired (Java) StreamAddressSpace
+     */
+    public static StreamAddressSpaceMsg getStreamAddressSpaceMsg(StreamAddressSpace addressSpace) throws IOException {
+        StreamAddressSpaceMsg.Builder addressSpaceMsgBuilder = StreamAddressSpaceMsg.newBuilder();
+        addressSpaceMsgBuilder.setTrimMark(addressSpace.getTrimMark());
+
+        try (ByteString.Output bso = ByteString.newOutput()) {
+            try (DataOutputStream dos = new DataOutputStream(bso)) {
```

Comment:
For the sake of consistency, I'd say the best approach that I know, could be:
```
public static Optional<StreamAddressSpaceMsg> getStreamAddressSpaceMsg(StreamAddressSpace addressSpace) {
        
        Optional<ByteString> maybeAddressMap = Optional.empty();
        try (ByteString.Output bso = ByteString.newOutput()) {
            try (DataOutputStream dos = new DataOutputStream(bso)) {
                addressSpace.getAddressMap().serialize(dos);
                maybeAddressMap = Optional.of(bso.toByteString());
            }
        } catch (IOException ex) {
            String msg = "getStreamAddressSpaceMsg: error=[{}] while serializing roaring64NavigableMap";
            log.error(msg, ex.getCause(), ex);
        }
        
        return maybeAddressMap.map(addressMap ->{
            StreamAddressSpaceMsg.Builder addressSpaceMsgBuilder = StreamAddressSpaceMsg.newBuilder();
            addressSpaceMsgBuilder.setTrimMark(addressSpace.getTrimMark());
            addressSpaceMsgBuilder.setAddressMap(addressMap);
            return addressSpaceMsgBuilder.build();
        });
    }
```

BTW, who's using `getStreamAddressSpaceMsg`? I didn't find anything in Corfu which uses it

Reply:
This method in the API isn't used yet. It will be used by the SequencerClient/Server and LogUnitServer when the PRs replacing the implementation of these RPCs are ready to go.